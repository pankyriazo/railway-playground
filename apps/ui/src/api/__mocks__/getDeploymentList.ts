import { GetDeploymentListResponse } from "@/api/types";

export const mockGetDeploymentList = {
  data: {
    deployments: {
      edges: [
        {
          node: {
            id: "4b049de0-1ce0-4e81-8036-26b252e3285c",
            status: "SUCCESS",
            updatedAt: "2024-03-17T08:50:14.973Z",
            environmentId: "632da482-c5b1-4570-87af-9ea225c335cb",
          },
        },
        {
          node: {
            id: "38b68be4-f86d-4987-8c48-dee293fda404",
            status: "SUCCESS",
            updatedAt: "2024-03-16T19:41:25.747Z",
            environmentId: "1beb4373-6b1f-43cb-8150-0e7214acea82",
          },
        },
      ],
    },
  },
} as const satisfies GetDeploymentListResponse;

export const mockGetDeploymentListError = {
  data: null,
  errors: [
    {
      message: "Unauthorized",
    },
  ],
} as const satisfies GetDeploymentListResponse;
