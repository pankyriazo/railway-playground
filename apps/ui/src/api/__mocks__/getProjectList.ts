import { GetProjectListResponse } from "@/api/types";

export const mockGetProjectList = {
  data: {
    me: {
      projects: {
        edges: [
          {
            node: {
              id: "a20a0d44-a6de-4804-acf1-58d436570fd7",
              name: "nauseating-cup",
              services: {
                edges: [
                  {
                    node: {
                      id: "0e62e7bf-2008-4452-bd32-e57cec3bb9a2",
                      name: "fascinated-motion",
                    },
                  },
                  {
                    node: {
                      id: "7ab42902-aae8-47b7-9976-81496dc61834",
                      name: "Postgres",
                    },
                  },
                ],
              },
              plugins: {
                edges: [],
              },
              environments: {
                edges: [
                  {
                    node: {
                      id: "1beb4373-6b1f-43cb-8150-0e7214acea82",
                      name: "production",
                    },
                  },
                  {
                    node: {
                      id: "632da482-c5b1-4570-87af-9ea225c335cb",
                      name: "staging",
                    },
                  },
                ],
              },
            },
          },
        ],
      },
    },
  },
} as const satisfies GetProjectListResponse;

export const mockGetProjectListError = {
  data: null,
  errors: [
    {
      message: "Unauthorized",
    },
  ],
} as const satisfies GetProjectListResponse;
