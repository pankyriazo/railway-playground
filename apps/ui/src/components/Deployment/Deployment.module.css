.container {
  background-color: var(--status-color__bg);
  border-radius: var(--border-radius);
  color: var(--status-color__text);
  cursor: default;
  display: flex;
  flex-direction: row;
  font-size: var(--font-size-xs);
  font-weight: var(--font-weight-medium);
  padding: var(--spacing-sm);
  row-gap: var(--spacing-xs);
  transition: background-color var(--transition);
}
.container:hover {
  background-color: var(--status-color__bg--hover);
}

.success {
  --status-color__bg: var(--color-success);
  --status-color__bg--hover: var(--color-success--hover);
  --status-color__text: var(--color-text--inverse);
}

.failed {
  --status-color__bg: var(--color-error);
  --status-color__bg--hover: var(--color-error--hover);
  --status-color__text: var(--color-text--inverse);
}

.pending {
  --status-color__bg: var(--color-warning);
  --status-color__bg--hover: var(--color-warning--hover);
  --status-color__text: var(--color-text);
}

.default {
  --status-color__bg: var(--color-disabled);
  --status-color__bg--hover: var(--color-disabled--hover);
  --status-color__text: var(--color-text--inverse);
}
