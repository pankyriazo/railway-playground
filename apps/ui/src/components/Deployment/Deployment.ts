import { Deployment as DeploymentType } from "@monorepo/types";
import { StatefulComponent } from "@/types";

import styles from "./Deployment.module.css";

class Deployment extends StatefulComponent<HTMLLIElement> {
  private _deployment: DeploymentType;

  constructor({ deployment }: { deployment: DeploymentType }) {
    super();

    this._element = document.createElement("li");
    this._element.id = `deployment-${deployment.id}`;
    this._deployment = deployment;

    this.render();
  }

  override render() {
    if (!this._element) {
      return;
    }

    this._element.innerHTML = "";
    this._element.dataset.testid = "deployment";
    this._element.className = styles.container!;
    if (
      this._deployment.status === "FAILED" ||
      this._deployment.status === "CRASHED"
    ) {
      this._element.dataset.status = "failed";
      this._element.classList.add(styles.failed!);
    } else if (this._deployment.status === "SUCCESS") {
      this._element.dataset.status = "success";
      this._element.classList.add(styles.success!);
    } else if (
      this._deployment.status === "DEPLOYING" ||
      this._deployment.status === "BUILDING"
    ) {
      this._element.dataset.status = "pending";
      this._element.classList.add(styles.pending!);
    } else {
      this._element.dataset.status = "default";
      this._element.classList.add(styles.default!);
    }

    const updatedAt = new Date(this._deployment.updatedAt).toLocaleString();
    const updatedAtElement = document.createElement("time");
    updatedAtElement.dataset.testid = "deployment__updated-at";
    updatedAtElement.className = styles.updatedAt!;
    updatedAtElement.dateTime = this._deployment.updatedAt;
    updatedAtElement.textContent = updatedAt;
    this._element.appendChild(updatedAtElement);
  }
}

export default Deployment;
