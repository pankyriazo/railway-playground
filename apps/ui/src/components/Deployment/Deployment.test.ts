import { render } from "@/test-helpers";
import { mockGetDeploymentList } from "@/api/__mocks__/getDeploymentList";
import Deployment from "./Deployment";

describe("Deployment", () => {
  const deployment = mockGetDeploymentList.data.deployments.edges[0].node;
  const statuses = {
    FAILED: "failed",
    CRASHED: "failed",
    SUCCESS: "success",
    DEPLOYING: "pending",
    BUILDING: "pending",
  } as const;

  let dateSpy: jest.SpyInstance;

  beforeAll(() => {
    dateSpy = jest
      .spyOn(Date.prototype, "toLocaleString")
      .mockReturnValue("1/1/2024, 10:50:14 AM");
  });

  afterAll(() => {
    dateSpy.mockRestore();
  });

  it("should render the deployment", () => {
    render(Deployment, { deployment });

    const deploymentElement = document.querySelector(
      "[data-testid=deployment]"
    );

    expect(deploymentElement).toMatchSnapshot();
  });

  Object.keys(statuses).forEach((value) => {
    describe(`when the deployment's status is '${value}'`, () => {
      const status = statuses[value as keyof typeof statuses];

      it(`should render the deployment as ${status}`, () => {
        render(Deployment, { deployment: { ...deployment, status: value } });

        const deploymentElement = document.querySelector(
          "[data-testid=deployment]"
        );

        expect(
          deploymentElement!.attributes.getNamedItem("data-status")!.value
        ).toBe(status);
      });
    });
  });

  describe("when the deployment's status is unknown", () => {
    it("should render the deployment as default", () => {
      render(Deployment, { deployment: { ...deployment, status: "UNKNOWN" } });

      const deploymentElement = document.querySelector(
        "[data-testid=deployment]"
      );

      expect(
        deploymentElement!.attributes.getNamedItem("data-status")!.value
      ).toBe("default");
    });
  });
});
