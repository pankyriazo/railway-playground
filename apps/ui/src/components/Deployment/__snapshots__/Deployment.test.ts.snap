// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deployment should render the deployment 1`] = `
<li
  class="container success"
  data-status="success"
  data-testid="deployment"
  id="deployment-4b049de0-1ce0-4e81-8036-26b252e3285c"
>
  <time
    class="updatedAt"
    data-testid="deployment__updated-at"
    datetime="2024-03-17T08:50:14.973Z"
  >
    1/1/2024, 10:50:14 AM
  </time>
</li>
`;
